### PR TITLE
feat: (qe) Telemetry capturing for PDP/Data proxy

### PR DESCRIPTION
**Pull request by miguelff**
Extracted from https://github.com/prisma/prisma-engines/pull/3505 and to be released after the preparatory refactoring in https://github.com/prisma/prisma-engines/pull/3617
 
Tracked in https://github.com/prisma/client-planning/issues/158
Sibling to https://github.com/prisma/prisma/pull/16748
Closes https://github.com/prisma/client-planning/issues/138
Closes https://github.com/prisma/client-planning/issues/144
Closes https://github.com/prisma/client-planning/issues/145

Implements traces and log capturing in concordance with the [(private link) contract specification](https://www.notion.so/prismaio/Data-Proxy-Logs-Scale-Team-Spec-184242eb7271476d82113c59c761bed9#4ec77c2aa9004015ac2d392c7458dfb5)

The technical design is best explained throw the interaction diagram below:

diagram

In the diagram, you will see objects whose lifetime is static. The boxes for those have a double
width margin. These are:

  - The `server` itself
  - The  global `TRACER`, which handles `log!` and `span!` and uses the global `PROCESSOR` to
   process the data constituting a trace `Span`s and log `Event`s
  - The global `PROCESSOR`, which manages the `Storage` set of data structures, holding logs,
   traces (and capture settings) per request.

Then, through the request lifecycle, different objects are created and dropped:

  - When a request comes in, its headers are processed and a [`Settings`] object is built, this
   object determines, for the request, how logging and tracing are going to be captured: if only
   traces, logs, or both, and which log levels are going to be captured.
  - Based on the settings, a new `Capturer` is created; a capturer is nothing but an exporter
   wrapped to start capturing / fetch the captures for this particular request.

Then the capturing process works in this way:
  
  - The server receives a query **[1]**
  - It grabs the HTTP headers and builds a `Capture` object **[2]**, which is configured with the settings
    denoted by the `X-capture-telemetry`
  - Now the server tells the `Capturer` to start capturing all the logs and traces occurring on
   the request **[3]** (denoted by a `trace_id`) The `trace_id` is either carried on the `traceparent`
   header or implicitly created on the first span of the request. To _start capturing_ implies
   creating for the `trace_id` in two different data structures: `logs` and `traces`; and storing
   the settings for selecting the Spans and Event to capture **[4]**.
  - The server dispatches the request and _Somewhere_ else in the code, it is processed **[5]**.
  - There the code logs events and emits traces asynchronously, as part of the processing **[6]**
  - Traces and Logs arrive at the `TRACER`, and get hydrated in the `PROCESSOR` **[7]**,
   which writes them in the shard corresponding to the current `trace_id`, into the
   `logs` and `traces` storage **[8]**. The settings previously stored under the `trace_id`
   key, are used to pick which events and spans are going to be captured based on their level.
  - When the code that dispatches the request is done it returns a `PrismaResponse` to the
   server **[9]**.
  - Then the server asks the `PROCESSOR` to fetch the captures **[10]**
  - And right after that, it fetches the captures from the `Storage` **[11]**. At that time, although
   that's not represented in the diagram, the captures are deleted from the storage, thus
   freeing any memory used for capturing during the request
  - Finally, the server sets the `logs` and `traces` extensions in the `PrismaResponse`**[12]**,
   it serializes the extended response in json format and returns it as an HTTP Response
   blob **[13]**.

### CODE REVIEW

File: query-engine/core/src/telemetry/capturing/capturer.rs [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+}
+
+impl Processor {
+    pub fn new() -> Self {
+        Self {
+            storage: Default::default(),
+        }
+    }
+
+    pub(self) async fn start_capturing(&self, trace_id: TraceId, settings: Settings) {
```
Comment (tomhoule) [inline query-engine/core/src/telemetry/capturing/capturer.rs:68]: I think `pub(self) fn` is equivalent to `fn`, that's the default visibility.

**Review by tomhoule — APPROVED** [review]
With all the context and future plans we discussed, let's merge this.

File: query-engine/core/src/interactive_transactions/mod.rs [hunk 2]
```diff
@@ -76,6 +76,63 @@ impl Display for TxId {
...
+}
+
+impl TxId {
+    pub fn as_traceparent(&self) -> String {
+        let trace_id = opentelemetry::trace::TraceId::from(self.clone());
+        format!("00-{}-0000000000000001-01", trace_id)
+    }
+}
+
+impl From<TxId> for opentelemetry::trace::TraceId {
```
[the remark this answers is not in the document]
Comment (miguelff) [reply query-engine/core/src/interactive_transactions/mod.rs:106]: Got it, makes sense. It's not obvious by looking at the code what the code does.

File: query-engine/core/src/telemetry/capturing/settings.rs [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    pub fn is_enabled(&self) -> bool {
+        self.traces_enabled() || self.logs_enabled()
+    }
+
+    #[inline(always)]
+    pub fn traces_enabled(&self) -> bool {
+        self.include_traces
+    }
+
+    #[inline(always)]
```
[the remark this answers is not in the document]
Comment (miguelff) [reply query-engine/core/src/telemetry/capturing/settings.rs:65]: This is a leftover after experimenting and trying to learn some compiler heuristics. Will remove.

File: query-engine/core/src/interactive_transactions/error.rs [hunk 4]
```diff
@@ -17,3 +22,27 @@ pub enum TransactionError {
...
+}
+
+impl From<CoreError> for ExtendedTransactionUserFacingError {
+    fn from(error: CoreError) -> Self {
+        ExtendedTransactionUserFacingError {
+            user_facing_error: error.into(),
+            extensions: Default::default(),
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (miguelff) [reply query-engine/core/src/interactive_transactions/error.rs:48]: Yes, you are.

File: query-engine/core/src/telemetry/capturing/capturer.rs (hunk 1 above)
Comment (miguelff) [reply query-engine/core/src/telemetry/capturing/capturer.rs:68]: noted, thank you. Will remove.

**Comment by miguelff** [comment]
https://github.com/prisma/prisma-engines/pull/3618/commits/80442d539bafcc15d2ff6e40b9fe600db84bf67e fixed the problem with `BEGIN` statements not being captured at the beginning of the request.

`curl  -X POST http://localhost:57581/transactions/start -H 'Content-Type: application/json' -H 'X-capture-telemetry: query,tracing' -d '{"max_wait":2000,"timeout":5000}'`

```json
{
  "id": "cldensyuh0002rb9nzo10w4iy",
  "extensions": {
    "logs": [
      {
        "span_id": "563b074839c5c57e",
        "level": "query",
        "timestamp": [
          1674832125,
          565891000
        ],
        "attributes": {
          "params": "[]",
          "query": "SELECT 1",
          "duration_ms": "12"
        }
      },
      {
        "span_id": "3b3817ccec15d3e9",
        "level": "query",
        "timestamp": [
          1674832125,
          572383000
        ],
        "attributes": {
          "query": "BEGIN",
          "duration_ms": "4",
          "params": "[]"
        }
      }
    ],
    "traces": [
      {
        "trace_id": "73797568303030327a6f313077346979",
        "span_id": "563b074839c5c57e",
        "parent_span_id": "fc6e12b498c57dce",
        "name": "prisma:engine:db_query",
        "start_time": [
          1674832125,
          553089000
        ],
        "end_time": [
          1674832125,
          566096000
        ],
        "attributes": {
          "db.statement": "SELECT 1"
        }
      },
      {
        "trace_id": "73797568303030327a6f313077346979",
        "span_id": "fc6e12b498c57dce",
        "parent_span_id": "6925d3c29b00efb1",
        "name": "prisma:engine:connection",
        "start_time": [
          1674832125,
          552560000
        ],
        "end_time": [
          1674832125,
          567526000
        ],
        "attributes": {
          "db.type": "postgres"
        }
      },
      {
        "trace_id": "73797568303030327a6f313077346979",
        "span_id": "3b3817ccec15d3e9",
        "parent_span_id": "6925d3c29b00efb1",
        "name": "prisma:engine:db_query",
        "start_time": [
          1674832125,
          567792000
        ],
        "end_time": [
          1674832125,
          572568000
        ],
        "attributes": {
          "db.statement": "BEGIN"
        }
      }
    ]
  }
}
```

File: query-engine/core/src/telemetry/capturing/capturer.rs (hunk 1 above)
Comment (miguelff) [reply query-engine/core/src/telemetry/capturing/capturer.rs]: Fixed in https://github.com/prisma/prisma-engines/pull/3618/commits/9a545b0a838ce42ddc08107d8e62180492f07888

File: query-engine/core/src/telemetry/capturing/settings.rs [hunk 5]
```diff
@@ -0,0 +1,88 @@
+use std::{collections::HashSet, time};
+
+static VALID_VALUES: &[&str] = &["error", "warn", "info", "debug", "trace", "query", "tracing"];
+static DEFAULT_TTL: time::Duration = time::Duration::from_secs(1800); // 30 minutes
+
+#[derive(Debug, Clone, Default)]
+pub struct Settings {
+    // only capture log events from the specified log levels, the special level "query", which does not
+    // exist in the engine logging infrastructure, is shimed from any event describing a query, regardless
+    // of its level.
```
[the remark this answers is not in the document]
Comment (miguelff) [reply query-engine/core/src/telemetry/capturing/settings.rs]: Fixed in https://github.com/prisma/prisma-engines/pull/3618/commits/9a545b0a838ce42ddc08107d8e62180492f07888

File: query-engine/core/src/interactive_transactions/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (miguelff) [reply query-engine/core/src/interactive_transactions/mod.rs]: Improved in https://github.com/prisma/prisma-engines/pull/3618/commits/a2c2b6170901d988dfa39b251100825b4d08ea53

**Pull request merged**